File: java/Problem3.java
// Problem3.java --- 
// Filename: Problem3.java
// Description: 
// Author: amritendu
// Email:  amritoit@gmail.com
// Organization:  IIT Madras
// Created: Sun Feb 28 10:00:37 2016 (+0530)
// Last-Updated: Sun Feb 28 10:37:25 2016 (+0530)
//           By: amritendu
//     Update #: 32
// 

public class Problem3{

    public static void main(String[] args){
        
        if(args[0]==null)System.out.println("Please pass the number to me: java <className> <number> ");
        System.out.println(greatestPrimeFactor(Long.parseLong(args[0])));
   
    }

    private static long greatestPrimeFactor(long n){
        
        int limit=(int)Math.sqrt(n)+1;
        int i=3;

        if(n==4)return 2;
        if(n==1)return 0;

        for(i=3;i<=limit;i=i+2){
            
            while(isEven(n)){
                n=n/2;
                if(n==1)return 2;
            }


            while(n%i==0)
                 n=n/i;

            if(n==1)break;
        }
        
        if(n==1)return i;
        else return n;
        
    }

    private static boolean isEven(long n){
        
        return (n & 1)==1?false:true;
    }
}

File: python/problem-3.py
# problem-3.py ---
# Filename: problem-3.py
# Description: problem: https://projecteuler.net/problem=3
# Author: Amritendu Mondal
# Email:  amritoit@gmail.com
# Organization:  IIT Madras
# Created: Thu Jul 25 22:43:31 2019 (+0530)
# Last-Updated:
#           By:
#     Update #: 21
#

# solution: bf approach will work as computation is fast.


import time


def compute(lower_limit, upper_limit):
    a = max(i*j
            for i in range(upper_limit, lower_limit, -1)
            for j in range(i, lower_limit, -1) if str(i*j) == str(i*j)[::-1])
    return a


start_time = time.time()
print("compute result =", compute(100, 999),
      ", time taken - ", (time.time() - start_time), " seconds")
assert compute(100, 999) == 906609

print("all test case passed")
